src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -2358,12 +2358,13 @@ private CorInfoHelpFunc getBoxHelper(CORINFO_CLASS_STRUCT_* cls)
         {
             var type = HandleToObject(cls);
 
-            // we shouldn't allow boxing of types that contains stack pointers
-            // csc and vbc already disallow it.
+            if (type.IsNullable)
+                return CorInfoHelpFunc.CORINFO_HELP_BOX_NULLABLE;
+
             if (type.IsByRefLike)
-                ThrowHelper.ThrowInvalidProgramException(ExceptionStringID.InvalidProgramSpecific, MethodBeingCompiled);
+                return CorInfoHelpFunc.CORINFO_HELP_BOX_SLOW;
```

Comment:
Just curious - why do we need a BOX_SLOW helper here? (And a runtime helper that can handle byref-like types at all).

Shouldn't this be a new helper `CORINFO_HELP_BOX_INVALID` that does nothing but throws InvalidProgram at runtime?

Would we ever be in a situation where we need a slow helper that can either box or throw? I would think such helper would only be needed if we have universal shared code (a "`__Canon`" type that could be a byref-like valuetype at runtime). We don't have universal shared code in CoreCLR and __Canon can only be a reference type.

Reply:
The BOX_SLOW is the JIT helper that always checks state and basically represents the JIT is naive case. I could add yet another helper for this that just handles all INVALID cases too. I'm not sure that is any better than this approach though. Is the argument that all box failure paths should be detectable at JIT time and we can determine which case then? I've not looked into it, but what about the interpreter?

Reply:
From what I saw it looks like we only hand out the SLOW helper when we already know it will throw. If instead if having a SLOW helper we had an INVALID helper that RyuJIT understands is always throwing, it could potentially remove unreachable code after the helper. If it's just a show helper that may or may not throw, it must assume it might return.

Reply:
I'm not saying we would do the unreachable optimization now - just as a structural thing.

Reply:
@dotnet/jit-contrib Is there prior art in the JIT for indicating a helper will "always throw"? I'm asking because @MichalStrehovsky's suggestion is compelling but adding a new helper implementation that will "always throw" seems a bit premature if there no existing concept. I agree the JIT could use this but unless the JIT is already aware of the concept it seems adding this type of helper now is a bit premature.

Reply:
Yes, there is the `GTF_CALL_M_DOES_NOT_RETURN` flag. Looks like there already are a few helpers we create that we known will throw exceptions using this concept, in particular for unsupported intrinsics. You can use `gtNewMustThrowException` to create a helper call with this flag.

Reply:
It might be better to create a helper `CORINFO_HELP_THROW_INVALIDPROGRAM` and use the above function to create a call to that. There are already other kinds of helpers throwing other exceptions.

Reply:
Thanks @jakobbotsch. That makes this decision easy. @MichalStrehovsky I'll update this logic as suggested.

Reply:
Well now that I look into this, I am less convinced. Creating either a `CORINFO_HELP_THROW_INVALIDPROGRAM` or `CORINFO_HELP_BOX_INVALID` results in odd statements to determine eliding recognized IL sequences in `Compiler::impBoxPatternMatch`.

```c++
if (boxHelper == CORINFO_HELP_BOX || boxHelper == CORINFO_HELP_THROWINVALIDPROGRAM)
```

or

```c++
if (boxHelper == CORINFO_HELP_BOX || boxHelper == CORINFO_HELP_BOX_INVALID)
```
I personally don't find these very clear and in this case we are now pushing even more policy into the JIT as opposed to letting the VM handle how to fail boxing a ByRefLike. Unless there is a strong case for this potential optimization, which I don't see because the JIT is typically going to inline the box, I am going to leave it as-is.

@MichalStrehovsky @jakobbotsch @jkotas ?

Reply:
I am not super familiar with this handling, @AndyAyersMS might know more. But if I understand correctly `impBoxPatternMatch` tries to optimize common boxes away, and is something we should not be doing in any case if the box helper might throw. If I am right then I would change `impImportBlockCode` to avoid calling that function for `CORINFO_HELP_THROWINVALIDPROGRAM` and fix the else case in `impImportAndPushBox` for the throwing case.

Note that you will need to make sure you spill `exprToBox` to retain side effects of the struct computation, as there are no args to the throwing helper -- the easy way to do that is with `impSpillSideEffects`, but you will need to reorder the code a bit to make sure we do not call `impPopStack` before that.

Reply:
> If I am right then I would change impImportBlockCode to avoid calling that function for CORINFO_HELP_THROWINVALIDPROGRAM and fix the else case in impImportAndPushBox for the throwing case.

This is by design for boxing with the ByRefLike generic support - see https://github.com/dotnet/runtime/pull/67129. We need to handle cases where a `box` instruction is created for a `T` that might throw if the branch is taken but isn't in practice.

Reply:
Ah okay, sorry, I have not realized that there are common patterns involving `box` that we still want to support. Then I do think it is a little weird to import a throw helper here and use it to determine this.

Since there is a special contract here for just by-ref like types, would it make sense to just check in the JIT if this is a byref-like type (see e.g. `verIsByRefLike` for a place we do that) and use that to handle the common idioms, and otherwise create the throwing helper? That way we do not involve `getBoxHelper` at all.

Reply:
Maybe you can add a new bool `bool* pInvalidProgram` arg to `getBoxHelper` to allow it to communicate to the JIT that it should throw InvalidProgramException.

Also, `impImportAndPushBox ` is implemented as opportunistic optimization today. It needs revisiting to turn it into a functional guarantee. 

For example, the `box` ; `br_true/false` combination will be optimized only if the tree to box does not have side-effects. vs. https://github.com/dotnet/runtime/pull/67129 says that it will be always optimized and treated as a single unit.

Reply:
> Since there is a special contract here for just by-ref like types,

But that is the point of how this was designed, that doesn't exist. The JIT doesn't care what the resolve type is and it defers to the VM to handle it in the best way it sees fit via the slow path.

> Maybe you can add a new bool bool* pInvalidProgram arg to getBoxHelper to allow it to communicate to the JIT that it should throw InvalidProgramException.

I don't see why we should let the JIT make this decision.

> Also, impImportAndPushBox is implemented as opportunistic optimization today. 

Do you mean `impBoxPatternMatch`?

> For example, the box ; br_true/false combination will be optimized only if the tree to box does not have side-effects.

Right. Let me look at that again.

Reply:
> But that is the point of how this was designed, that doesn't exist. The JIT doesn't care what the resolve type is and it defers to the VM to handle it in the best way it sees fit via the slow path.

The JIT is required to optimize several IL patterns for by-ref like types. That is the special contract I am talking about, and why I think this should be checked directly in the JIT. It cannot call the helper provided by the runtime even though it sees a box instruction in these cases, so it does care what the resolve type is.